### DISCUSSION TITLE
My git fetch command is broken after updating from 3.1.20 to 3.1.30

### DISCUSSION THREAD

**Discussion by gpxricky** [gitpython-developers/GitPython#1540] [Q&A] [answered]
Hi,
in gitpython version 3.1.20 I was using the following construct:

```
# The list is filled with information from an external source like a configuration file, current content is just an example
fetch_options = [ "--depth", "1", "tag", tag_to_fetch ] 
remote.fetch(fetch_options)
```

It worked fine, although I wasn't sure if I used the function properly. After updating to version 3.1.30 the following git error is raised: 'fatal: couldn't find remote ref --depth'. This is because the git cmd command changed to:

```
git fetch -v -- origin --depth 1 tag MY_TEST_TAG
```

Which is obviously wrong now. How do I need to change my fetch call in order to fetch a tag with some additional options for the command that may be read from a text file?

Many thanks and best regards,
Markus

**Comment by stsewd** [comment]
Hi, similar to https://github.com/gitpython-developers/GitPython/issues/1528, you were relying on a security vulnerability that allowed positional arguments to be interpreted as options (the first argument from fetch is meant to be a list of refs)

https://github.com/gitpython-developers/GitPython/blob/90c81a56dffe77fc08f863769d35762ca66240b0/git/remote.py#L954

You need to pass additional options as kwargs.

**Comment by gpxricky** [comment]
@stsewd Thanks for the information. So at the end, if I need to remain backwards compatible and can't change the input file, then the only solution is to change my command to

```
fetch_options = [ "-v", "origin", "--depth", "1", "tag", tag_to_fetch ] 
repo.git.fetch(fetch_options)
```

? I think converting the list to a dictionary is quite complicated as an option may be used with short name or long name, may be written as "--depth=1" or "--depth 1" and may contain multiple values, a single value or no value.

Best regards,
Markus

**Reply by Byron** [reply]
What about `remote.fetch(depth=1, tag=tag_to_fetch)`?

**Comment by gpxricky** [comment, marked as answer]
@Byron Many thanks. I think it is clear now for me how to use the "new" fetch function. Unfortunately, I don't know which options are provided to my function as they are read from an input file and I can't change the input file due to compatibility issues and other problems. Therefore my function still looks as follows:

```
def fetch_tag(remote, tag, fetch_options):
    remote.fetch(?)
```

In the past I was using the `fetch` method of a `Remote`, but I think I need to change that to use git directly:

```
def fetch_tag(remote, tag, fetch_options):
    remote.repo.git.fetch(["-v", remote.name, "tag", tag] + fetch_options)
```

This is more or less fine for me, although using the official method feels somehow better.

Another option would be to convert my list of `fetch_options` to a dictionary, but I think it is not worth the effort and it may also get a little bit complicated.

**Reply by Byron** [reply]
If both ways of doing things are semantically the same for you, I don't think there is the need to avoid `repo.git.fetch(…)` at all.
